
## Promises

Asenkron kodlar ile baş edebilmenin başka bir yolu da *promises*lerdir.

Önceki bölümde gördüğünüz üzere, callbackler ile  bir fonksiyonun çalışması bittiğinde, işleme alınacak fonksiyonu çağırıyorduk.

Örneğin:

```js
doSomething(result => {
  console.log(result)
})
```
`doSomething()` fonksiyonu sona erdiğinde, parametre olarak gönderilen fonksiyon çağrılır.

```js
const doSomething = callback => {
  //do things
  //do things
  const result = /* .. */
  callback(result)
}
```
Buradaki ana problem fonksiyondan gelecek sonucu, programın geri kalanında kullanmak istememizle ortaya çıkmaktadır. Bu durumda bütün kodumuzu callback'in içine yerleştirmemiz gerekmektedir. İki yada üç tane callback yapmamız durumunda , "callback cehennemi" adını verdiğimiz sisteme girintili birçok fonksiyon katmanı ile dahil olurduk.

```js
doSomething(result => {
  doSomethingElse(anotherResult => {
    doSomethingElseAgain(yetAnotherResult => {
      console.log(result)
    })
  }) 
})
```
Bu sorunla başa çıkmanın bir yolu Promiselerdir.

Bunu yapmak yerine:

```js
doSomething(result => {
  console.log(result)
})
```
Promise temelli fonksiyonu bu şekilde çağırabiliriz:

```js
doSomething()
  .then(result => {
    console.log(result)
  })
```

Öncelikle fonksiyonu çağrırız, fonksiyon bittiği zamansa `then()`adını verdiğimiz fonksiyonu çağırıyoruz.

Satır girintisi önemli olmasa da okunabilirlik için bu stili sıkça kullanacaksınız.

`catch()` metodu, hataları yakalamak için sıkça başvurulan bir yöntemdir:

```js
doSomething()
  .then(result => {
    console.log(result)
  })
  .catch(error => {
    console.log(error)
  })
```
Bu söz dizimini kullanmak için, Promises API'sini kullanarak, `doSomething()` fonksiyonun uygulanmasını özel bir hale getirmeliyiz.

Normal fonksiyon tanımlamak yerine:

```js
const doSomething = () => {
  
}
```

Bir promise nesnesi tanımlıyoruz.

```js
const doSomething = new Promise()
```

ve fonksiyonunu promise constructorına gönderiyoruz.

```js
const doSomething = new Promise(() => {

})
```

Bu fonksiyon, iki parametre almaktadır. İlki promisi resolve eden, ikincisiyse reject eden fonksiyonlardır. 
```js
const doSomething = new Promise(
  (resolve, reject) => {
    
})
```

Bir promise'i resolve etmek demek onu başarılı bir şekilde tamamlamak tamamlamak anlamına gelmektedir(Herhangi bir kullanımda `then()` metodu çağrısından gelen sonuçtur).
Bir promise'i reject etmek demek bir hatayla sonlanması anlamına gelir(Bahsedilen hata, herhangi bir kullanımda `catch()` metodu çağrısından gelen sonuçtur.)

Bunu yapmanın yolu:

```js
const doSomething = new Promise(
  (resolve, reject) => {
    //some code
    const success = /* ... */
    if (success) {
      resolve('ok')
    } else {
      reject('this error occurred')
    }
  }
)
```
Resolve ve reject fonksiyonlara istediğimiz türde parametreler yollayabiliriz.
